### ISSUE TITLE
aPLib format changes for improving decompression on 8-bit micros.

### ISSUE THREAD

**Issue by jbrandwood** [emmanuel-marty/apultra#1]
Hi Emmanuel, thank you for this project, it is so wonderful to see a pure-C packer for the aPLib format the can be included in other projects (particularly development for old comupters/consoles)!

You may wish to check out my [aplpak](https://github.com/jbrandwood/aplpak) repository which includes a 6502 aPLib decompressor that executes approximately 2.5x the speed of Peter Ferrie's 65C02 code that is included in Jorgen Ibsen's aPLib distriution (at the cost of 27 bytes - 256 vs 229).

My aPLib code is decompressing at approx 75 cycles-per-byte, which compares favorably to Peter Ferrie's 6502 LZSA2 decompressor, which is decompressing at 77 cycles-per-byte with the same test file (gpl-3.0.txt). 

Are you open to making some (optional) alterations to the aPLib format to improve the decompression speed a little bit more?

One obvious change (IMHO) would be to nibble-pack the 4-bit offset in the copy-one-byte-from-within-15-bytes "111 nnnn" encoding.

Another change would be to use seperate bit-buffers for the main 2-or-3-bit compression-type code, and the Elias gamma-codes, because the gamma-codes are always read 2-bits-per-loop, and so the routine that reads them would only need to do a single check for reading a new byte if the 2-bits were always 2-bit-aligned in the bit-buffer.

Does that explanation make sense?

Are you open to tweaking the format a little?

**Comment by emmanuel-marty** [comment]
Hey John,

Thanks for the kind words! I am definitely open to having an extended format to speed up decompression on 8-bit micros (as you probably guessed from LZSA, it is my center of interest for these packers). 

Here is a patch that implements your suggested changes, essentially creating three different bit buffers, one for reading 1 bit, one for reading 2 bits, and one for reading 4. Here's also the corresponding exe, you'll need to pass the -e command line flag to (de)compress with the extended format, otherwise it will behave as the regular aPLib. Since on 6502 you can't keep the bit shifter in a register, using more than one wouldn't penalize you, so skipping the test for every other bit in the gamma2 decoder and for 3 out of 4 bits for the short match offsets should provide a nice boost for sure. The compressed file may be 1-2 bytes larger as the extra bit buffers are padded out, but if we get a decompression boost that's definitely fine.

Let me know if these changes are what you had in mind, and if so I'm curious to see the results. If it all looks good I'll commit the changes to github. I'll be happy to add both your original 6502 aPLib decoder and the extended format one to this repo as well, if you like.

Happy to make more changes to the extended format as well if more ideas come to mind.

Unrelated, but if you can take a look at LZSA2 in particular, I'd be very interested in a faster version of the 6502 decompressor if some of your aPLib optimizations can apply to it. Happy to publish a third, faster/larger version or to merge it with a #define. (I'm guessing that skipping the cost of the JSR/RET to get source bytes and use indirect Y addressing instead would be the main optimization)

[apultra_win64_1.0.9_extformat.zip](https://github.com/emmanuel-marty/apultra/files/3994305/apultra_win64_1.0.9_extformat.zip)

[ext_format_patch.txt](https://github.com/emmanuel-marty/apultra/files/3994304/ext_format_patch.txt)

**Comment by jbrandwood** [comment]
Thank you!

I'll test the changes and let you know what the speed improvment turns out to be.

Yes, the main speedup in my code comes from removing all of the JSR/RET overhead that is in Peter Ferrie's code.

While it is possible to do some self-modifying code in the inner-loop of the copy, or to use a subroutine in zero-page ... the JSR/RET overhead overwhelms any advantages, since the majority of literal & match lengths in aPLib are only a few bytes long.


One thing that I'd like to ask you to consider is to implement a "window" (a.k.a. maximum offset) into the compressor.

That is because old machines such as the PC Engine, NES, Gameboy, ColecoVision, TI/99, etc had large amounts of port-accessed (i.e. not random access) VRAM compared to main RAM, and so compressors for those machines often decompress directly to VRAM, and use a small "window" buffer in main RAM, just like a lot of traditional LZSS implementations.

Most games that I have looked at use either 256 byte or 4KB window sizes, but any power-of-two is possible.

On the 6502 and Z80/Gameboy, a 256-byte window is really fast to implement by just incrementing the bottom byte of the address (or X & Y register).

This hurts overall compression ratio, but it's the best that you can do when you only have a few KB of main RAM.

The aPLib format is a good fit for something like this since the hi-byte of the offset is already gamma-coded.

Just something else for you to think about! ;-)

**Comment by emmanuel-marty** [comment]
Looking forward to the speed improvement results!

Once I checked in that code (if it turns out to be useful), I can definitely add a configurable window size next. It mostly just needs to be exposed to the public library functions and to the command-line frontend.

**Comment by jbrandwood** [comment]
I have done some testing, and I'm afraid that the results are OK, but not as compelling as I had hoped.

The TLDR result is that with your changes, plus another two minor changes, I can get an 11% speedup in decompression, at the cost of an extra 18 bytes of code (270 bytes, since I reduced 4 bytes from my previous decompressor).

That makes the "enhanced" format approximately 50% slower than LZSA2 for only a 5% improvement in compression ratio, which seems like a poor tradeoff unless you're desperate to fit things into a limited amount of memory (i.e. a cartridge or floppy disc).

The additional changes that I have added are to switch the continue/stop flag in the gamma code (just to save a byte on the 6502), and to split long gamma codes so that you only have to rotate an 8-bit value instead of a 16-bit value (since a LOT of time is spent reading the gamma codes).

This may be worth adding to APULTRA for the few people that *really* need the compression-ratio of aPLib, but I could easily understand it if you didn't want to.

I am attaching the latest 6502 decompressor (for both original and enhanced formats), and the patch-from-your-patch with the minor changes that I made.

FYI, the I tested with both an Atari progam binary, and the gpl-3.0txt, and the results were similar. Here are the results for gpl-3.0.txt, since it is an easy file to find (tested on the HuC6280) ...

2622137 cycles to decompress gpl-3.0.txt apultra STANDARD format (252 bytes of code) (35147 bytes decompressed, 74.6 cycles per byte)

2345286 cycles to decompress gpl-3.0.txt apultra ENHANCED format (270 bytes of of code, 18 bytes longer) (35147 bytes decompressed, 66.7 cycles per byte) saves approx 11%

[aplib_6502.txt](https://github.com/emmanuel-marty/apultra/files/4004183/aplib_6502.txt)

[apu-enhanced.txt](https://github.com/emmanuel-marty/apultra/files/4004190/apu-enhanced.txt)

P.S. To split out the speed improvements ... the nibble packing saves approximately 4%, using a seperate buffer for gamma code pairs saves approximately 3%, and splitting long gamma codes saves another 3%.

**Comment by emmanuel-marty** [comment]
Thanks! As you are the only 'consumer' of this extended format so far, 11% on aPLib is pretty solid, and aPLib can use all the speed it can get on 8-bitters, I merged the support for the format, it doesn't harm anything. You're welcome to use both LZSA2 and aPLib in your toolchain (they both use an identical libdivsufsort, the rest of the libs should combine cleanly). 

I will attach a patch for a window size option shortly, that will limit the max offset to use for backreferences, so that you can decompress to port RAM as well

**Comment by emmanuel-marty** [comment]
Here is a tentative patch to implement the window size setting with -w <bytes>
[apultra_window.txt](https://github.com/emmanuel-marty/apultra/files/4005963/apultra_window.txt)

Let me know if it does what you need, and I'll subsequently commit the changes, or amend it as needed

**Comment by jbrandwood** [comment]
Thank you for the "window" patch, it seems to be working perfectly! :-)

May I suggest that you change the patch itself to remove the use of the min() function, since that isn't standard across all compilers.

Instead something like this, which mimics the line of code above the change ...

_const int nBlockSize = nMaxWindowSize ? ((nDefaultBlockSize < nMaxWindowSize / 2) ? nDefaultBlockSize : nMaxWindowSize / 2) : nDefaultBlockSize;_

As expected, the window hurts compression, but the gamma-coded offsets help to reduce the pain. Using the window and decompressing to a port is about 50% slower than normal decompression, but that's also expected for those machines that just don't have enough main RAM (a PCE HuCard game for example).

I'll be interested to see what happens to LZSA2 if you wish to apply a window to that, since once-again, its variable-length offset encoding provides a couple of natural window sizes!

**Comment by emmanuel-marty** [comment]
Merged
